File: C#/Binary Search/875. Koko Eating Bananas.cs
/* 875. Koko Eating Bananas
Difficulty: Medium
Submission Link: https://leetcode.com/problems/koko-eating-bananas/submissions/1376888525/ */

public class Solution { 
    public int MinEatingSpeed(int[] piles, int h) {
        int low=1, high=piles.Max(); 

        while(low<=high){
            int mid=(low+high)/2;
            double val=0;

            foreach (var pile in piles) {
                val += Math.Ceiling((double)pile/mid);
            }
             
            if(val<=h){ 
                high=mid-1;
            }
            else{
                low=mid+1;
            }
        }
        return low;
    }
}

File: Python3/Binary Search/875. Koko Eating Bananas.py
""" 875. Koko Eating Bananas
Difficulty: Medium
Submission Link: https://leetcode.com/problems/koko-eating-bananas/submissions/1376886153/ """

class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        # k small, h large. 
        
        def counter(k):
            count = 0
            for pile in piles:
                count += math.ceil(pile/k)
            return count
        start, end = 1, max(piles)
        while start < end:
            mid = (start+end) //2 
            temp_h = counter(mid)
            if temp_h <= h:
                end = mid
            else:
                start = mid +1
        
        return start